transport/src/main/java/io/netty/channel/SubChannel.java

```diff
@@ -0,0 +1,520 @@
+package io.netty.channel;
+
+import java.net.SocketAddress;
+import java.util.concurrent.RejectedExecutionException;
+
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.RecvByteBufAllocator.Handle;
+import io.netty.util.Attribute;
+import io.netty.util.AttributeKey;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+public class SubChannel implements Channel {
```

Comment:
Do we think SubChannel is the best name choice?  Some other ideas: Nested, Inner, Child.

Would it be possible (in a spearate CR) to implement Netty's HTTP/2 child/stream channels with this class, or are they too fundamentally different?  Would be a great sign of a solid implementation, and it'd consequentially receive more attention over time.

Also please add some Javadoc on the class' primary purpose, as well as call out how it behaves identical to parent vs how it behaves differently.

(I'm kind of leaning towards "Child" given that's the terminology Netty has already used with HTTP/2, and it fits well with your chosen use of "parent".)

Reply:
I'm OK with Child if that is the consensus. 

My SubChannel implementation was based heavily on the ```DefaultHttp2StreamChannel``` implementation in ```codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java```, but quite simplified to delegate functionality to the parent channel for various features, such as ```attr()```, etc that would not be appropriate for a more generic use case. 

It could be made more generic, and reimplemented as something like ```MultiplexedChannel``` with a special case where there is only a single channel multiplexed on the parent (being effectively this ```SubChannel```). And then ```DefaultHttp2StreamChannel``` could be implemented on top of it. That said, ```DefaultHttp2StreamChannel``` has quite a lot of functionality in it that diverges from my goal for ```SubChannel```, particularly maintaining its own queue of Objects to be processed. I'm not quite sure of the rationale for this, it seems like another place where buffering may interfere with readabilty/writability of the channel. But I have not yet dug deeply into HTTP/2!